[Feature] Support of masked materials in instance segmentation (UE4)

tdegraaff opened the pull request:
#### Description

Currently, the instance segmentation replaces all materials on all meshes simply using an opaque material, which can be colorized with an RGB encoding of the type and the object ID. Secondary SceneProxies of all meshes are then rendered with these tagged materials resulting in the instance segmentation output. The current solution has the drawback, that masked materials are simply ignored. Thus, partial transparent surfaces in the RGB image appear fully opaque in the instance segmentation, resulting in very coarse annotations, which also do not match to the semantic segmentation.

This PR addresses this issue by introducing additional logic, that tackles masked materials for the instance segmentation. The following images show a before/after comparison of the semantic channel in the instance segmentation.

|         Town01_Opt          |          Town10HD_Opt           |
| :-------------------------: | :-----------------------------: |
| ![Town01-1](https://github.com/user-attachments/assets/179be03a-4d01-440a-963a-10a0cc24dea4) | ![Town10HD-1](https://github.com/user-attachments/assets/442b4d31-cf30-4a0e-885b-2b266b42eba7) |
| ![Town01-2](https://github.com/user-attachments/assets/e14299fc-4c30-4816-90b5-de25766dcd23) | ![Town10HD-2](https://github.com/user-attachments/assets/8dd39ba9-adb4-448a-9d9f-fca1139a16a7) |

In general, the instance segmentation on vegetation and fences is now as fine-grained as the semantic segmentation. Note, that the main problem with the sky and the lane markings in Town01 (and Town02) are fixed. Also, trees in Town10HD in the background (with LOD4) were previously invisible (since it uses a very special material) and are now visible. The only remaining difference between the semantic segmentation and the instance segmentation is, that the semantic segmentation looks through translucent surfaces, while the instance segmentation does not.

#### Implementation

* The core of this PR is the `UTaggedMaterialsRegistry`, which is now responsible to handle tagged materials for the `UTaggedComponents`
  * It holds a mapping from original materials to modified materials, which are duplicates of the originals, but introduce some additional expressions into the material graph, so that they render with the tag color
  * This keeps the relevant aspects of the original material intact (especially `OpacityMask` and `WorldPositionOffset`), so that the duplicated material behaves like the original, except that it renders with the tag color
* If a `UTaggedComponent's` parent uses a masked material, a tag-injected corresponding material is requested from the registry
  * Otherwise, the default opaque tagged material is used
* When playing in editor (PIE), these tagged materials are created and compiled on demand
* For the packaged version, a commandlet creates a `UTaggedMaterialsRegistry` for the current package, which then will be cooked and copied to the package

#### Where has this been tested?

* **Platform(s):** Windows 11, Ubuntu 18.04
* **Python version(s):** 3.8
* **Unreal Engine version(s):** 4.26
* Tested in Editor (Windows) and Packaged (Windows + Linux)

#### Possible Drawbacks

IMO, very small drawbacks, but for completion:
* Packaging takes some additional time, due to additionally generating and cooking the `TaggedMaterialsRegistry`
* The resulting package(s) grow in size, due to the additional `TaggedMaterialsRegistry`
  * For the "Carla" package (which is the largest), additionally 44MB (Windows) or 25MB (Linux) are needed, respectively
  * Other packages (e.g. "AdditionalMaps" or "Town07_Opt") only consume additionally ~1MB
* When playing a map in the editor (PIE) the first time, creating and compiling the required materials slows down the startup
  * Since these are stored, all further in-game sessions in the editor of this map are not affected anymore
* When playing in the editor but as "Standalone", runtime modification and compiling of materials is not possible
  * Warnings are logged and the instance segmentation falls back to the usual fully-opaque tagged material (i.e. same behavior before this PR)
  * If the map was previously played in editor, it also works in standalone

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/carla-simulator/carla/8795)
<!-- Reviewable:end -->

Blyron commented:
We will need to discuss this PR as we are aiming to reduce building and cooking times. I cannot guaranteed this PR is gonna be merged.

tdegraaff commented:
I think the increase of building time from this PR is negligible small. But I totally agree, that we should aim to not massively increase the cooking times. And when packing all packages, there is indeed some noteworthy overhead coming from this PR.

Until the end of this week, I will try to:
- measure the additional cooking time coming from the current implementation of this PR (so that we know, what we are talking about) and
- find potential in reducing this overhead (I already have some ideas), then measure again.

I hope that we find some solutions so that we can get this merged :)

Blyron commented:
That would be awesome. Thank you for your effort!
We have some times about the current process;
Build Engine: 1.5h
Build CARLA: 15min
Download Content: 1h
Whole package: 3h

Blyron commented:
Hello any news on cooking times?

tdegraaff commented:
I am sorry for the long delay, realizing and testing of my initial ideas was not as trivial as I assumed.

#### Benchmarking

I now have some results from measuring the packaging process. Testing was done on a Windows 11 PC with an i5-14600KF with 64GB RAM. I deleted the `Saved` and `DerivedDataCache` in `CarlaUE4` and also the `DerivedDataCache` in my Unreal Engine build folder, to start packaging from scratch.

| Package                         | Overall duration | This PR's share | Increase |
| :------------------------------ | ---------------: | --------------: | -------: |
| Overall                         |   3h 35 min 24 s |     12 min 40 s |    6.25% |
| \|- Carla*                      |   1h 46 min 05 s |      4 min 14 s |    4.16% |
| \|- All Add. Maps**             |   1h 39 min 19 s |      8 min 26 s |    9.28% |
| &nbsp;&nbsp; \|- AdditionalMaps |       5 min 14 s |      1 min 14 s |   30.83% |
| &nbsp;&nbsp; \|- Town06_Opt     |       3 min 41 s |            59 s |   36.42% |
| &nbsp;&nbsp; \|- Town07_Opt     |       4 min 13 s |            59 s |   30.41% |
| &nbsp;&nbsp; \|- Town11         |      28 min 15 s |      1 min 07 s |    4.12% |
| &nbsp;&nbsp; \|- Town12         |      16 min 41 s |      1 min 20 s |    8.69% |
| &nbsp;&nbsp; \|- Town13         |      33 min 07 s |      1 min 21 s |    4.25% |
| &nbsp;&nbsp; \|- Town15         |       4 min 42 s |      1 min 26 s |   43.88% |

*Carla: `make package`
**All Add. Maps: `make package ARGS="--packages=AdditionalMaps,Town06_Opt,Town07_Opt,Town11,Town12,Town13,Town15 --target-archive=AdditionalMaps --clean-intermediate"`

So, there is really some major additional overhead of ~13 minutes on the packaging time currently coming from this PR. I am glad that you pointed me to take a deeper look into this, I was not aware, that the overhead is that tremendous. Getting to zero is obviously not possible, since additional stuff needs to be cooked, but we can definitely do better. 👇

#### What I did so far

I now did the following:
1) Added time measuring and logging to the Package.* scripts, making it easy to inspect the required time for different steps in the process (I used this to derive the above results). I think it is useful to always have this in the logs.
2) Took a deeper look into Unreal's cooking implementation and tested with some cooking flags.
   * My main hope was on `-NeverCookDirectories` (which is currently not usable in UE4, so I backported that from UE5). But this actually only prevented to write the assets in the given directories, they were compiled nevertheless, bringing no time reduction.
   * I added `-NoShaderCompile` to the call of the `GenerateTaggedMaterialsRegistry` commandlet, at least avoiding on-the-fly compiling already here.
3) Changed the `GenerateTaggedMaterialsRegistry` commandlet to not be called per package but for all packages at once, avoiding lots of overhead from initializing the Unreal Editor per call and also some previously redundant scanning of files in the commandlet itself.
   * It was cleaner and easier to call it individually, but that is just to much waste of time.
   * This change brings gains when packaging multiple packages at once, i.e. reducing the time for packaging All Add. Maps from 8.5 minutes down to 2 minutes.

#### My "last" idea and todos

1) I will now take a look into the duplicated materials, since I have the impression, that too much stuff is compiled. I assume this comes from unnecessarily duplicated parts in the material graph, so I will try to only duplicate those parts, that are actually required.
   * If this goes well, I will come back and report the final packaging times.
2) I have to test all these changes again on Linux

tdegraaff commented:
Ok, got it to work with not duplicating full materials but only selected parts of the material graph. The gain is smaller than I have hoped, but it is there:
`make package` now additionally costs 3 min 20 s (4 min 14 s previously)
`make package ARGS="--packages=AdditionalMaps,Town06_Opt,Town07_Opt,Town11,Town12,Town13,Town15 --target-archive=AdditionalMaps --clean-intermediate"` now additionally costs 1 min 30 s (8 min 26 s previously).

Additional packaging time thus increases now by 5 min (13 min previously). Would you be ok with that?
I am still testing on Linux and would refactor last code. I hope to push the changes tomorrow.

Blyron commented:
Okay. We will have a look

Blyron commented:
Are you gonna update it more or it is okay to go?

tdegraaff commented:
I am sorry for the long delay.
Now everything is updated with optimizations that reduce the cooking times (as described above).
I tested it on Windows and Linux now. I also rebased the branch onto the current ue4-dev branch.

Blyron commented:
Perfect. Thanks! I will test as soon as possible

jorge-kabuto reviewed (approved):
> Reviewed 19 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/carla-simulator/carla/8795)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @tdegraaff)



<!-- Sent from Reviewable.io -->
